File: internal/core/archive.go
package core

import (
	"archive/tar"
	"archive/zip"
	"bytes"
	"compress/gzip"
	"fmt"
	"io"
	"regexp"
)

const (
	// MaxFileSize is the maximum size of a single file during extraction (100 MB).
	MaxFileSize = 100 * 1024 * 1024
	// MaxTotalSize is the maximum total size of all extracted files (1 GB).
	MaxTotalSize = 1024 * 1024 * 1024
)

// ExtractedFile represents a file extracted from an archive.
type ExtractedFile struct {
	Name string
	Data []byte
}

// ExtractTarGz extracts files from tar.gz data in memory.
// This is used by both CLI and WASM for in-memory extraction.
// For file-based extraction, use the manifest package.
func ExtractTarGz(tarGzData []byte) ([]ExtractedFile, error) {
	return ExtractTarGzReader(bytes.NewReader(tarGzData))
}

// ExtractTarGzReader extracts files from a tar.gz reader.
func ExtractTarGzReader(r io.Reader) ([]ExtractedFile, error) {
	gzr, err := gzip.NewReader(r)
	if err != nil {
		return nil, fmt.Errorf("creating gzip reader: %w", err)
	}
	defer gzr.Close()

	tr := tar.NewReader(gzr)
	var files []ExtractedFile
	var totalSize int64

	// Regex to detect path traversal
	pathTraversal := regexp.MustCompile(`(^|/)\.\.(/|$)`)

	for {
		header, err := tr.Next()
		if err == io.EOF {
			break
		}
		if err != nil {
			return nil, fmt.Errorf("reading tar: %w", err)
		}

		// Security: reject path traversal
		if pathTraversal.MatchString(header.Name) {
			return nil, fmt.Errorf("archive contains invalid path: %s", header.Name)
		}

		// Skip directories, symlinks, and other special files
		if header.Typeflag != tar.TypeReg {
			continue
		}

		// Security: enforce file size limits
		if header.Size > MaxFileSize {
			return nil, fmt.Errorf("file %s exceeds maximum allowed size (%d bytes)", header.Name, MaxFileSize)
		}
		totalSize += header.Size
		if totalSize > MaxTotalSize {
			return nil, fmt.Errorf("archive exceeds maximum total size (%d bytes)", MaxTotalSize)
		}

		// Use LimitReader for additional safety
		limitedReader := io.LimitReader(tr, MaxFileSize)
		data, err := io.ReadAll(limitedReader)
		if err != nil {
			return nil, fmt.Errorf("reading file %s from archive: %w", header.Name, err)
		}

		files = append(files, ExtractedFile{
			Name: header.Name,
			Data: data,
		})
	}

	if len(files) == 0 {
		return nil, fmt.Errorf("empty archive")
	}

	return files, nil
}

// ExtractZip extracts files from ZIP data in memory.
// Same security checks (path traversal, size limits) as ExtractTarGz.
func ExtractZip(zipData []byte) ([]ExtractedFile, error) {
	r, err := zip.NewReader(bytes.NewReader(zipData), int64(len(zipData)))
	if err != nil {
		return nil, fmt.Errorf("opening zip: %w", err)
	}

	var files []ExtractedFile
	var totalSize int64

	// Regex to detect path traversal
	pathTraversal := regexp.MustCompile(`(^|/)\.\.(/|$)`)

	for _, f := range r.File {
		// Security: reject path traversal
		if pathTraversal.MatchString(f.Name) {
			return nil, fmt.Errorf("archive contains invalid path: %s", f.Name)
		}

		// Skip directories
		if f.FileInfo().IsDir() {
			continue
		}

		// Security: enforce file size limits
		if int64(f.UncompressedSize64) > MaxFileSize {
			return nil, fmt.Errorf("file %s exceeds maximum allowed size (%d bytes)", f.Name, MaxFileSize)
		}
		totalSize += int64(f.UncompressedSize64)
		if totalSize > MaxTotalSize {
			return nil, fmt.Errorf("archive exceeds maximum total size (%d bytes)", MaxTotalSize)
		}

		rc, err := f.Open()
		if err != nil {
			return nil, fmt.Errorf("opening zip entry %s: %w", f.Name, err)
		}

		limitedReader := io.LimitReader(rc, MaxFileSize+1)
		data, err := io.ReadAll(limitedReader)
		rc.Close()
		if err != nil {
			return nil, fmt.Errorf("reading file %s from archive: %w", f.Name, err)
		}
		if int64(len(data)) > MaxFileSize {
			return nil, fmt.Errorf("file %s exceeds maximum size during extraction", f.Name)
		}

		files = append(files, ExtractedFile{
			Name: f.Name,
			Data: data,
		})
	}

	if len(files) == 0 {
		return nil, fmt.Errorf("empty archive")
	}

	return files, nil
}

// ExtractArchive detects the archive format and extracts accordingly.
// ZIP archives start with PK\x03\x04, gzip with \x1f\x8b.
func ExtractArchive(data []byte) ([]ExtractedFile, error) {
	if len(data) < 2 {
		return nil, fmt.Errorf("archive too small to detect format")
	}

	// ZIP: starts with PK\x03\x04
	if data[0] == 0x50 && data[1] == 0x4B {
		return ExtractZip(data)
	}

	// gzip: starts with \x1f\x8b
	if data[0] == 0x1f && data[1] == 0x8b {
		return ExtractTarGz(data)
	}

	return nil, fmt.Errorf("unrecognized archive format (expected ZIP or tar.gz)")
}

File: internal/html/assets/src/crypto/archive.ts
// Archive extraction (tar.gz and ZIP) and tlock container detection
import { parseTar } from 'tarparser';
import { unzipSync } from 'fflate';
import type { TlockContainerMeta } from '../types';

export interface ExtractedFile {
  name: string;
  data: Uint8Array;
}

/**
 * Extract files from a tar.gz archive.
 */
export async function extractTarGz(data: Uint8Array): Promise<ExtractedFile[]> {
  const entries = await parseTar(data);

  return entries
    .filter(entry => entry.type === 'file')
    .map(entry => ({
      name: entry.name,
      data: entry.data,
    }));
}

/**
 * Extract files from a ZIP archive.
 */
export function extractZip(data: Uint8Array): ExtractedFile[] {
  const files = unzipSync(data);
  const result: ExtractedFile[] = [];

  for (const [name, fileData] of Object.entries(files)) {
    // Skip directories (they end with / and have empty data)
    if (name.endsWith('/')) continue;
    result.push({ name, data: fileData });
  }

  if (result.length === 0) {
    throw new Error('empty archive');
  }

  return result;
}

/**
 * Detect archive format and extract files.
 * ZIP archives start with PK (0x50 0x4B), gzip with 0x1F 0x8B.
 */
export async function extractArchive(data: Uint8Array): Promise<ExtractedFile[]> {
  if (data.length < 2) {
    throw new Error('archive too small to detect format');
  }

  // ZIP: starts with PK\x03\x04
  if (data[0] === 0x50 && data[1] === 0x4B) {
    return extractZip(data);
  }

  // gzip: starts with \x1f\x8b
  if (data[0] === 0x1f && data[1] === 0x8b) {
    return extractTarGz(data);
  }

  throw new Error('unrecognized archive format');
}

/**
 * Check if data is a tlock container ZIP (contains tlock.json).
 * Uses only fflate — no tlock.js dependency.
 */
export function isTlockContainer(data: Uint8Array): boolean {
  if (data.length < 4 || data[0] !== 0x50 || data[1] !== 0x4B || data[2] !== 0x03 || data[3] !== 0x04) {
    return false;
  }
  try {
    const files = unzipSync(data);
    return 'tlock.json' in files;
  } catch {
    return false;
  }
}

/**
 * Open a tlock container ZIP and return the metadata and tlock-encrypted ciphertext.
 * Uses only fflate — no tlock.js dependency.
 */
export function openTlockContainer(data: Uint8Array): { meta: TlockContainerMeta; ciphertext: Uint8Array } {
  const files = unzipSync(data);

  const metaData = files['tlock.json'];
  if (!metaData) {
    throw new Error('tlock container: missing tlock.json');
  }

  const ciphertext = files['manifest.tlock.age'];
  if (!ciphertext) {
    throw new Error('tlock container: missing manifest.tlock.age');
  }

  const meta: TlockContainerMeta = JSON.parse(new TextDecoder().decode(metaData));
  if (!meta.v) {
    throw new Error('tlock container: tlock.json missing version field');
  }

  return { meta, ciphertext };
}
